### PR TITLE
Feature/12 roles

### PR DESCRIPTION
**Pull request by mabiede**
- migrate search endpoints (experiment)
- add search location
- add rules page in settings
- add admin roles to the detail page
- add grant/revoke role form to admin edit page
- move the `Field` module to its own file
- add a signature for Lwt_result.Infix for the language server
- update yarn
- add tests for admin role assignments

### CODE REVIEW

File: pool/web/handler/admin_experiments.ml [hunk 1]
```diff
@@ -411,4 +385,22 @@ end = struct
       |> Guardian.validate_generic
     ;;
   end
+
+  let search =
```
Comment (timohuber) [inline pool/web/handler/admin_experiments.ml:389]: If there are multiple urls to search experiments, why not split this up in multiple middlewares? That would be easier to understand.

File: pool/web/utils/http_utils.ml [hunk 2]
```diff
@@ -215,17 +214,43 @@ let urlencoded_to_flash urlencoded =
...
+let find_in_urlencoded_base_opt = CCList.assoc_opt ~eq:CCString.equal
+
+let find_in_urlencoded_list_opt field
+  : (string * string list) list -> string list option
+  =
+  let open Pool_common.Message in
+  find_in_urlencoded_base_opt Field.(array_key field)
+;;
+
+let find_in_urlencoded_filtered_list field =
```
Comment (timohuber) [inline pool/web/utils/http_utils.ml:226]:
```suggestion
let find_in_urlencoded_list_opt field =
```
Wouldn't that be a better name?

File: pool/web/utils/http_utils.ml [hunk 3]
```diff
@@ -215,17 +214,43 @@ let urlencoded_to_flash urlencoded =
...
+  find_in_urlencoded_list_opt field %> CCOption.value ~default:[]
+;;
+
+let find_in_urlencoded_list field =
+  let open Pool_common.Message in
+  find_in_urlencoded_list_opt field %> CCOption.to_result (Invalid field)
```
Comment (timohuber) [inline pool/web/utils/http_utils.ml:232]:
```suggestion
  find_in_urlencoded_list_opt field %> CCOption.to_result (NotFound field)
```

File: pool/web/view/component/component.ml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-module Field = Pool_common.Message.Field
+module Cascade = Component_cascade
```
Comment (timohuber) [inline pool/web/view/component/component.ml:1]: What about naming it `Search`?

File: resources/admin/filter.js [hunk 5]
```diff
@@ -1,43 +1,10 @@
+import { addCloseListener, addInputListeners, csrfToken, destroySelected, icon, notifyUser } from "./utils.js";
```
Comment (timohuber) [inline resources/admin/filter.js:1]: Not sure if this bug is new, but if a filter contains a `Participation` predicate, a title is required.

If you want, I can have a look at the js part.

Comment (timohuber) [reply resources/admin/filter.js:1]: This was fixed with: https://github.com/uzh/pool/pull/133

File: pool/web/handler/admin_experiments.ml [hunk 6]
```diff
@@ -32,41 +24,16 @@ let find_all_with_role database_label role =
...
-            (fun actor ->
-              Guard.Persistence.validate
-                database_label
-                (read_validation_set id)
-                actor
-              ||> CCOption.(of_result %> is_some))
-            actor)
-        experiments
-    in
+    let%lwt filtered = Helpers.Guard.Filter.experiments context experiments in
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/web/handler/admin_experiments.ml]: Please create an issue to add pagination / search for that page.

File: pool/web/handler/admin_experiments.ml (hunk 1 above)
Comment (timohuber) [reply pool/web/handler/admin_experiments.ml:390]: As there is only one route, please cleanup the access middleware.

File: pool/web/handler/admin_experiments.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/web/handler/admin_experiments.ml]: pagination is done and I created an issue for the search routes

File: pool/web/handler/admin_experiments.ml (hunk 1 above)
Comment (mabiede) [reply pool/web/handler/admin_experiments.ml]: done

File: pool/app/admin/admin.ml [hunk 7]
```diff
@@ -1,11 +1,34 @@
...
+  Guard.Persistence.Actor.find_by_roles
+    ~ctx:(Pool_database.to_ctx pool)
+    ?exclude
+    roles
+  ||> map snd
+  ||> flatten
+  ||> uniq ~eq:Guard.Uuid.Actor.equal
+  ||> map CCFun.(Guard.Uuid.Actor.to_string %> Pool_common.Id.of_string)
+  >|> Repo.find_multiple pool
+;;
```
[the remark this answers is not in the document]
Comment (mabiede) [reply pool/app/admin/admin.ml]: thanks, done

**Pull request merged**